### PR TITLE
Upgrade Hadoop BAM for better interval filtering

### LINKED ISSUE #399: Upgrade HadoopBAM to use BAM index
Right now we use samtools for locally BAM reading since it can use the BAM index and directly access the reads of interest.

Hadoop BAM should support this now as well (https://github.com/HadoopGenomics/Hadoop-BAM/pull/59) and will allow us to only access a subset of the BAM as oppposed to filtering after loading when using the the `--loci` argument.

### PR DESCRIPTION
**Pull request by arahuja**
This brings back #433 and fixes #399 

@ryan-williams I think you tested switching the Java version on a single Spark submission, can we add something to the README on that? This PR will move the repository to Java 8 only

---
This change is [Reviewable](https://reviewable.io/reviews/hammerlab/guacamole/569)

### CODE REVIEW

**Comment by ryan-williams** [comment]
OK, sorry for the delay here @arahuja, this lgtm.

I just sent #578 that resolves the conflicts here and adds a couple of things.

I am agnostic about whether you incorporate any/all of it or just rebase this and merge it in yourself.

Lmk if you have any questions about what I'd proposed over on #578.

File: src/main/scala/org/hammerlab/guacamole/readsets/ReadSets.scala [hunk 1]
```diff
@@ -321,14 +320,15 @@ object ReadSets extends Logging {
...
     var result = reads
-    for {
-      overlapsLoci <- filters.overlapsLociOpt
-    } {
-      val contigLengths = getContigLengths(sequenceDictionary)
-      val loci = overlapsLoci.result(contigLengths)
-      val broadcastLoci = reads.sparkContext.broadcast(loci)
-      result = result.filter(_.asMappedRead.exists(broadcastLoci.value.intersects))
-    }
+    filters
```
Comment (ryan-williams) [inline src/main/scala/org/hammerlab/guacamole/readsets/ReadSets.scala]: I find the `for` to be cleaner, do you not?

Comment (arahuja) [inline src/main/scala/org/hammerlab/guacamole/readsets/ReadSets.scala]: No - not particularly?

**Comment by arahuja** [comment]
I re-based this PR, and then added your commit on top. 

I added one more commit to change a matching on `Option` to be a `fold` instead as I think matching on `Option` is not recommended.

I similarly changed a `for` on option to `foreach`.

**Comment by ryan-williams** [comment]
Sounds good.

> matching on Option is not recommended.

Interesting; any reason?

**Comment by arahuja** [comment]
>Interesting; any reason? 

Just boilerplate already built-in to `map`, `fold` or `foreach`

**Pull request merged**